### DISCUSSION TITLE
Secret Management

### DISCUSSION THREAD

**Discussion by johannes-b** [keptn/keptn#2829] [Q&A] [answered]
I want to start discussing the secret management for action-providers (aka. [Keptn-services](https://keptn.sh/docs/0.8.x/integrations/)) in Keptn. 

To get an idea about the surrounding use-case (which is mainly focused on executing a remediation action by an action-provider), please take a looked [here](https://github.com/keptn/keptn/discussions/2827#discussion-75341). The other thread explains what an `action-provider` is and why an `action-provider` needs access to credentials, such as an API endpoint and token; username and password, etc. 

### Status quo

With Keptn 0.7, an action-provider has direct access to a Kubernetes secret that holds the credentials the service needs for execution. 

Let's use an example here: 
* We have an action-provider that has the job of triggering an AnsibleTower runbook. Consequently, the action-provider needs to know the API endpoint of the AnsibleTower and the authentication token to interact with the API. This information is stored in a K8s secret. This secret is read when the action-providers performs its action. 
* Assuming this action-provider is also responsible for multiple projects or multiple stages, there would be multiple secrets holding the project/stage-specific credentials. For example, it is very likely that a production stage has another AnsibleTower instance than for example a hardening/QA/staging stage.  

The following picture should explain this status quo: 

![image](https://user-images.githubusercontent.com/729071/102607239-3e05d900-4128-11eb-8405-41dac5855cf8.png)

### User needs

The situation (problem) has now two faces: 

1. (_for a production setup_) Keptn users already have a secret management tool (e.g., [Vault](https://learn.hashicorp.com/vault), KeyPass, etc.) and want to re-use it for a production setup. In other words, they don't want to add the secrets to Keptn but rather connect Keptn with the secret management tool that is in place. This should be possible without changing the implementation of action-providers (Keptn-services).
2. (_for a demo scenario or POC_) Keptn users don't want to plug in their secret management tool, but rather use an out-of-the-box (and lightweight) solution offered by Keptn; this is required for a getting-started with Keptn scenario. 

### Proposed idea

The idea is now to introduce a *secret API* that can be used by action-providers to fetch the secrets they need. This should abstract away the actual implementation of the secret management tool, which can be either the Kubernetes based approach with K8s:Secrets, or a custom implementation (e.g., Vault). 

Consequently, the action-provider would not access the K8s directly, but talks with the API to get the credentials. The way this communication is authenticated, is discussed [here](https://github.com/keptn/keptn/discussions/2827#discussion-75341).

![image](https://user-images.githubusercontent.com/729071/102605926-54129a00-4126-11eb-9a5b-e83b15a84675.png)

*Open questions:*
* Which endpoints must be implemented by the *secret API*? 

### Feedback is very welcome

🙏 If you have an opinion on the proposed idea or if you need more details, please let us know. The idea is not set in stone and still subject to change.

**Comment by christian-kreuzberger-dtx** [comment]
**What does it feel like for the user?**

While I see the need of a technical discussion here, I also feel that we must define first how this feature feels like for the end-user, e.g.:

```
keptn create secret --project=xyz --service=foobar --stage=dev --fromFile=./secrets.yaml
```

Questions that arise from just putting this command as a proposal:

* Should secrets be defined per project? Per service? Per stage? Is there any inheritance between those?
* What is a secret? Is it a yaml file? Is it a key value pair?

Another question that comes to my find: How do we name secrets? E.g., dynatrace-service allows selecting a secret per stage:
```
dtCreds: dynatrace-production
```

So the above command will also need a name, e.g.:
```
keptn create secret dynatrace-production --project=xyz --service=foobar --stage=dev --fromFile=./secrets.yaml
```

**Reply by johannes-b** [reply]
Great input and concerns!

**For an end-user - using the Bridge - it should feel as follows:** (*Note:* The following mockup is still WIP) 

![image](https://user-images.githubusercontent.com/729071/110312245-204c5f80-8005-11eb-9ebf-bfd436a5269a.png)

This answers your questions as follows:

- Should secrets be defined per project? Per service? Per stage? Is there any inheritance between those? > ~~Yes, it should be possible to define secrets for different entities (project, stage, service), like we handle resources.~~ (I have to re-think my answer)
- What is a secret? Is it a yaml file? Is it a key-value pair? > A key-value map should be sufficient, even though a plain yaml would be the most generic approach and allows any kind of object. 

**Secret name:**
I agree with you, a secret needs a name.

**Comment by christian-kreuzberger-dtx** [comment]
**Should end-users be able to fetch a secret**?

Example use-cases:

* User wants to list all secrets (globally, per project, per service, per stage; e.g. `kubectl get secret -n ${PROJECT}-${STAGE}`
* User wants to inspect a certain secret (e.g., `kubectl get secret ...` - Note that this might not be possible with all secret providers)

**Comment by warber** [comment, marked as answer]
# Proposal (FINAL, with review comments incorporated)

The Secrets API shall be implemented as a separate service called `secret-service`.
The service itself does not store secrets but makes use of different types of *secret backends* which shall be configurable via the environment variable `BACKEND_TYPE`. As a first version only `BACKEND_TYPE=kubernetes` is supported.

In the beginning, we want to keep things simple, thus, the API basically consists of CRUD operations to manage secrets.

A secret consists of `key-value` pairs and is bound to a `scope`.
A scope is used to group secrets and can be bound to different access control lists.
Note that an API for managing scopes will not be implemented in this first version.

## Secrets API

### General
- Base Path `/secrets/v1`


### Create Secret
| Method | Path	  |
|--------|--------|
|`POST` | `/secret`|
```
POST http://<HOST>:<PORT>/secrets/v1/secret
{
  "name": "dbaccess",
  "scope": "keptn",
  "data": {
   "dbuser": "thor",
   "dbpass": "hammer"
  }
}
```

### Update Secret
| Method | Path	  |
|--------|--------|
|`PUT` | `/secret`|
```
POST http://<HOST>:<PORT>/secrets/v1/secret
{
  "name": "dbaccess",
  "scope": "keptn",
  "data": {
   "dbuser": "captain-america",
   "dbpass": "shield"
  }
}
```

Note that it is only possible to update the `whole` secret, and not individual key-value pairs.

### Delete Secret
| Method | Path	  |
|--------|--------|
|`DELETE` | `/secret`|

```
DELETE http://<HOST>:<PORT>/secrets/v1/secrets?name=dbaccess&scope=keptn
```

## Scope Authorization
Each `scope` shall be configurable of being able to read, create, update or delete a secret.
**Proposal**: 

```
scopes.yml

- scope: dynatrace
  capabilities:
  - dynatrace-read-secrets
    permissions:
    - read
  - dynatrace-manage-secrets
    permissions:
    - read
    - update
```

### Example usage (kubernetes backend):
`keptn create secret --name=dynatrace-token --scope=dynatrace --backend=kubernetes <...>`

Two Kubernetes roles will be created:
```
metadata:
  name: dynatrace-read-secrets
  namespace: keptn
  
[...]

rules:
  - verbs:
    - get
resources:
  - secrets
resourceNames: ["dynatrace-token"]
[...]

```

and

```
metadata:
  name: dynatrace-manage-secrets
  namespace: keptn
  
[...]

rules:
  - verbs:
    - get
    - update
resources:
  - secrets
resourceNames: ["dynatrace-token"]
[...]

```

A service using the secret needs to be deployed with a **binding** to the one or the other role.

**Reply by johannes-b** [reply]
The proposal looks very good! I have just one remark: Can we use singular on the endpoints? `/secrets` --> just `/secret` 

Or is there a reason for going with plural?

**Reply by warber** [reply]
I was just following best practices on REST APIs.

The controlPlane endpoint is using sigular: `GET /api/controlPlane/v1/project/{project}/stage/{stage}` wheras `GET /api/controlPlane/v1/projects/{project}/stages/{stage}` is I think more "accepted" to be best practice.

It becomes even more visible when you take a look at `GET /api/controlPlane/v1/project/{project}/stage` which is supposed to list all stages. However, this is not clear when looking at the endpoint. IMO, plural would definitely be better.

However, I think your point is, that we should be consistent with the already existing endpoints, right?
This is a valid argument. 

Is `/api/secrets/v1/secret` ok?

However, we should discuss if we switch to the plural notation in the near future for all endpoints.

**Reply by johannes-b** [reply]
Right, I'm concerned about being consistent with the endpoints already available. However, I like your suggestion of `/api/secrets/v1/secret` this makes sense to me.

**Comment by bacherfl** [comment]
> From an implementation point of view, for Kubernetes secrets, the scope would be mapped to a Kubernetes namespace in which the secret will be created 

Would it be better to create all secrets in the namespace where keptn is installed? This way we can limit the write/read secret permission to the namespace where keptn is installed, as opposed to granting Keptn read/write access for secrets in all namespaces

**Reply by agrimmer** [reply]
I agree with @bacherfl. We can even consider not offering a GET endpoint for the first version because then the backend service does not require a role that allows reading all secrets in the namespace. 
However, I see the need for a PUT endpoint for updating secrets.

**Reply by agrimmer** [reply]
Would it be possible to implement `scope` and `access control list` with Kubernetes objects like `Role` and `RoleBinding`?
Can we do this exercise?

**Reply by warber** [reply]
@bacherfl I also agree with not binding the scope to a Kubernetes namespace. Changed the proposal

**Reply by warber** [reply]
@agrimmer added PUT endpoint

**Reply by warber** [reply]
@agrimmer 
> Would it be possible to implement `scope` and `access control list` with Kubernetes objects like `Role` and `RoleBinding`?

I was thinking in that direction and I think that should work. The only thing needed is to ensure that particular `scopes` are bound to particular access tokens in the `secret-service`.
